Index: CLAM/src/Processing/Analysis/FFT_base.hxx
===================================================================
--- CLAM/src/Processing/Analysis/FFT_base.hxx        (revisión: 12087)
+++ CLAM/src/Processing/Analysis/FFT_base.hxx        (working copy)
@@ -106,6 +106,16 @@
 	 */
 	virtual bool Do(const Audio& in, Spectrum &out)= 0;
 
+		//PyHack
+		bool Do(const ProcessingData& in, ProcessingData &out) {
+			try {
+					Do( dynamic_cast<const Audio&>(in), dynamic_cast<Spectrum&>(out) );
+			}
+			catch(...){
+					return false;
+			}
+		}
+
 	// Input/Output configuration methods
 
 	/** FFT non-supervised mode SetPrototypes function */

Index: CLAM/src/Processing/Analysis/FFT_ooura.hxx
===================================================================
--- CLAM/src/Processing/Analysis/FFT_ooura.hxx	(revision 12081)
+++ CLAM/src/Processing/Analysis/FFT_ooura.hxx	(working copy)
@@ -95,7 +95,16 @@
 
 		bool Do(const Audio& in, Spectrum &out);
 
-		
+		//Python bindings hack
+ 		bool Do(const ProcessingData& in, ProcessingData &out) {
+ 			try {
+ 				Do( dynamic_cast<const Audio&>(in), dynamic_cast<Spectrum&>(out) );
+ 			}
+ 			catch(...){
+ 				return false;
+ 			}
+ 		}
+
 		bool MayDisableExecution() const {return true;}
 
 	};
Index: CLAM/src/Standard/Array.hxx
===================================================================
--- CLAM/src/Standard/Array.hxx	(revision 12081)
+++ CLAM/src/Standard/Array.hxx	(working copy)
@@ -206,6 +206,12 @@
 		new(&mpData[where]) T(elem);
 		mSize++;
 	}
+	void	EditElem(int where,const T& elem)
+	{
+		CLAM_DEBUG_ASSERT(where>=0,msgIndexOutOfRange);
+		CLAM_DEBUG_ASSERT(where<mSize,msgIndexOutOfRange);
+		mpData[where] = elem;
+	}
 	void	DeleteElem(int where)
 	{
 		CLAM_ASSERT(OwnsMemory(),"Array::DeleteElem(): Resize requiered,"
