Index: generate_bindings.py
===================================================================
--- generate_bindings.py	(revisión: 13473)
+++ generate_bindings.py	(copia de trabajo)
@@ -125,7 +125,8 @@
 cflags_list = os.popen( 'PKG_CONFIG_PATH=%s/lib/pkgconfig pkg-config --cflags-only-other %s'%(clam_prefix," ".join(clam_modules))  ).read()[:-1].split(' ')
 define_list = [ item[2:] for item in cflags_list if item[:2]=='-D' ]
 
-flags_list = ""
+flags_list = "--gccxml-compiler gcc-4.2" #Force gccxml 4.2 version (needed in systems with greater versions as default, i.e.: ubuntu jaunty)
+#flags_list = ""
 
 # Creating an instance of class that will help you to expose your declarations
 mb = module_builder.module_builder_t (
@@ -133,7 +134,8 @@
 	working_directory = r".",
 	include_paths = clam_include_path,
 	define_symbols = define_list,
-	indexing_suite_version = 2,
+	indexing_suite_version = 1,
+#	indexing_suite_version = 2, # indexing suite v2 seems to has problems with recent versions?
 	cflags = flags_list,
 )
 
